_posts/2018-12-16-simple-front-end.md

```diff
@@ -0,0 +1,96 @@
+---
+layout: post
+author: nathan
+title: Web app front ends for people who don't want to have to continually re-learn front end
+---
+
+> If you're a javscript/front end developer who follows the latest sophisticated frameworks, or a lead with a team of rockstar front end developers looking to build the best app your designer can conceive of; then this isn't for you.
+
+Not every tool and technology is right for every task or situation. These are the driving goals for the approach outlined below
+
+1. Simplicity
+	* Minimize the number of dependencies / imports
+	* Minimize prerequisite knowledge
+	* Minimize code complexity
+2. Stability
+	* Dependencies / imports used should still be maintained years from now, and should have very few breaking changes in that time
+3. Scalability
+	* In performance, the number of active users should not impact the front end itself
+	* In functionality (if not design), we can implement an interface to work with a modern back-end API (we could implement amazon, but not lucidchart)
+
+With that said, some context on the range of front end's out there
+
+### GLAMP
+The Good old LinuxApacheMysqlPhp stack. This isn't strictly LAMP, but generally anything which renders dynamic data into HTML on the back end and returns it to the browser.
+
+* Synchronous
+	* What's shown to the user is driven by a user interaction resulting in loading new pages 
+* Server side rendering
+#### Pros
+* Can be very performant on low end devices since there is little or no javascript running
+* Pages with data from many sources may load faster because of better network performance between the front end server and the data sources compared to the browser and the back end / data sources
+#### Cons
+* Performance scales poorly, because pages cannot be cached
+* New content requires a page reload
+
+### SPA
+The modern Single Page Application is largely script driven. 
+
+* Asynchronous 
+	* What's shown on to the user is driven by user interaction or programmatic event resulting in scripts changing some content on the current page
+* Client side rendering
+	* Little or no HTML is delivered to the browser. Rather, javscript creates all the elements on the page.
```

Comment:
Important to note here that some modern JS frameworks do server-side rendering. So the pre-rendered components with app initial state are pre-rendered in the server and then the client code hooks on to these. These setups typically require a bundling system (e.g. webpack) in place and a bit of configuration.